File: algorithms/code/hackerrank/count_luck/count_luck.rs
// https://www.hackerrank.com/challenges/count-luck

use std::collections::VecDeque;

use crate::hackerrank::tester::{read_input, write_and_check_output};

#[derive(Clone, Copy)]
struct Here {
    row: i32,
    col: i32,
}

fn look_around(forest: &[Vec<char>], at: &Here) -> Vec<Here> {
    let deltas = [(0, 1), (0, -1), (1, 0), (-1, 0)];
    let mut ways: Vec<Here> = Vec::new();
    for delta in deltas.iter() {
        let dr = at.row + delta.0;
        let dc = at.col + delta.1;
        if dr < 0 || forest.len() <= dr as usize {
            continue;
        }
        if dc < 0 || forest[dr as usize].len() <= dc as usize {
            continue;
        }
        if 'X' == forest[dr as usize][dc as usize] {
            continue;
        }
        ways.push(Here { row: dr, col: dc });
    }
    ways
}

pub fn count_luck(matrix: &[String], k: i32) -> String {
    let mut queue: VecDeque<Here> = VecDeque::new();
    let mut counts: Vec<i32> = vec![0];
    let mut forest: Vec<Vec<char>> = matrix.iter().map(|s| s.chars().collect()).collect();
    for (r, row) in forest.iter().enumerate() {
        for (c, _) in row.iter().enumerate() {
            if 'M' == row[c] {
                queue.push_back(Here {
                    row: r as i32,
                    col: c as i32,
                });
            }
        }
    }
    while let Some(at) = queue.pop_front() {
        let count = counts[0];
        counts.remove(0);
        if '*' == forest[at.row as usize][at.col as usize] {
            if count == k {
                return "Impressed".to_string();
            }
            return "Oops!".to_string();
        }
        forest[at.row as usize][at.col as usize] = 'X';
        let ways = look_around(&forest, &at);
        for way in ways.iter() {
            queue.push_back(*way);
            if ways.len() > 1 {
                counts.push(count + 1);
            } else {
                counts.push(count);
            }
        }
    }
    "Oops!".to_string()
}

pub fn main_tested(name: &str) {
    let origin = file!();
    let input = read_input(origin, name);
    let tests = input[0][0].parse::<usize>().unwrap();
    let mut results: Vec<String> = Vec::new();
    let mut offset = 1;
    for _ in 0..tests {
        let n = input[offset][0].parse::<usize>().unwrap();
        let matrix: Vec<String> = input[offset + 1..offset + 1 + n]
            .iter()
            .map(|line| line[0].clone())
            .collect();
        let k = input[offset + 1 + n][0].parse::<i32>().unwrap();
        offset += 1 + n + 1;
        results.push(count_luck(&matrix, k));
    }
    write_and_check_output(origin, name, &results);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        main_tested("_example");
    }

    #[test]
    fn test_empty() {
        let result = count_luck(&[], 0);
        assert_eq!(result, "Oops!");
    }
}

File: algorithms/code/hackerrank/count_luck/count_luck.go
// Package countluck implements https://www.hackerrank.com/challenges/count-luck
package countluck

type here struct {
	row int
	col int
}

func lookAround(forest [][]rune, at here) []here {
	deltas := [][]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}
	ways := make([]here, 0)
	for _, delta := range deltas {
		dr := at.row + delta[0]
		dc := at.col + delta[1]
		if dr < 0 || len(forest) <= dr {
			continue
		}
		if dc < 0 || len(forest[dr]) <= dc {
			continue
		}
		if forest[dr][dc] == 'X' {
			continue
		}
		ways = append(ways, here{dr, dc})
	}
	return ways
}

const oops = "Oops!"

// CountLuck - implements the solution to the problem
func CountLuck(matrix []string, k int32) string {
	queue := make([]here, 0)
	counts := make([]int32, 0)
	counts = append(counts, 0)
	forest := make([][]rune, len(matrix))
	for r := range matrix {
		forest[r] = []rune(matrix[r])
		for c := 0; c < len(forest[r]); c++ {
			if forest[r][c] == 'M' {
				queue = append(queue, here{r, c})
			}
		}
	}
	for len(queue) != 0 {
		at := queue[0]
		queue = queue[1:]
		count := counts[0]
		counts = counts[1:]
		if forest[at.row][at.col] == '*' {
			if count == k {
				return "Impressed"
			}
			return oops
		}
		forest[at.row][at.col] = 'X'
		ways := lookAround(forest, at)
		for _, way := range ways {
			queue = append(queue, way)
			if len(ways) > 1 {
				counts = append(counts, count+1)
			} else {
				counts = append(counts, count)
			}
		}
	}
	return oops
}
